File: cortex/python/app/components/integration_hub.py
"""
Integration Hub Component

This module implements a hub for managing external integrations and services.
It provides a centralized way to configure, initialize, and interact with
external APIs, services, and data sources.
"""

import asyncio
import importlib
import inspect
import pkgutil
from abc import ABC, abstractmethod
from enum import Enum
from typing import Any, Dict, List, Optional, Set, Type, Union

from pydantic import BaseModel, Field

from app.config import settings
from app.utils.logger import get_contextual_logger

# Configure logger
logger = get_contextual_logger("components.integration_hub")


class IntegrationStatus(str, Enum):
    """Integration status enum"""

    NOT_CONFIGURED = "not_configured"
    CONFIGURED = "configured"
    INITIALIZING = "initializing"
    ACTIVE = "active"
    ERROR = "error"
    DISABLED = "disabled"


class IntegrationConfig(BaseModel):
    """Base model for integration configuration"""

    enabled: bool = True
    name: str
    type: str
    config: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class IntegrationInfo(BaseModel):
    """Information about an integration"""

    name: str
    type: str
    description: str
    status: IntegrationStatus
    capabilities: List[str] = Field(default_factory=list)
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class Integration(ABC):
    """Abstract base class for all integrations"""

    def __init__(self, config: IntegrationConfig):
        """Initialize the integration with config"""
        self.config = config
        self.status = IntegrationStatus.CONFIGURED
        self._initialized = False

        # Create integration-specific logger
        self.logger = get_contextual_logger(f"integration.{config.type}.{config.name}")

        self.logger.info(f"Integration {config.name} ({config.type}) configured")

    @abstractmethod
    async def initialize(self) -> bool:
        """
        Initialize the integration

        Returns:
            True if initialization was successful, False otherwise
        """
        pass

    @abstractmethod
    async def shutdown(self) -> None:
        """Shutdown the integration and release resources"""
        pass

    @abstractmethod
    def get_info(self) -> IntegrationInfo:
        """
        Get information about this integration

        Returns:
            Integration information
        """
        pass

    @property
    def initialized(self) -> bool:
        """Check if the integration is initialized"""
        return self._initialized

    @property
    def enabled(self) -> bool:
        """Check if the integration is enabled"""
        return self.config.enabled

    @property
    def name(self) -> str:
        """Get the integration name"""
        return self.config.name


class APIIntegration(Integration):
    """Base class for API-based integrations"""

    def __init__(self, config: IntegrationConfig):
        """Initialize the API integration"""
        super().__init__(config)

        # API-specific configuration
        self.base_url = self.config.config.get("base_url", "")
        self.api_key = self.config.config.get("api_key", "")
        self.timeout = self.config.config.get("timeout", 30)

    @abstractmethod
    async def test_connection(self) -> bool:
        """
        Test the API connection

        Returns:
            True if connection is successful, False otherwise
        """
        pass


class DatabaseIntegration(Integration):
    """Base class for database integrations"""

    def __init__(self, config: IntegrationConfig):
        """Initialize the database integration"""
        super().__init__(config)

        # Database-specific configuration
        self.connection_string = self.config.config.get("connection_string", "")
        self.pool_size = self.config.config.get("pool_size", 5)

    @abstractmethod
    async def test_connection(self) -> bool:
        """
        Test the database connection

        Returns:
            True if connection is successful, False otherwise
        """
        pass


class IntegrationHub:
    """
    Integration Hub for managing external integrations

    This class provides a central point for configuring, initializing,
    and accessing external integrations and services.
    """

    def __init__(self):
        """Initialize the integration hub"""
        self.integrations: Dict[str, Integration] = {}
        self.integration_classes: Dict[str, Type[Integration]] = {}

        # Register built-in integration types
        self._register_integration_types()

        logger.info("Integration hub initialized")

    def _register_integration_types(self) -> None:
        """Register built-in integration types"""
        # In a real implementation, this would dynamically discover and register
        # integration classes from a specific package/directory

        # For now, we'll just register some example integration types
        self.integration_classes = {
            "api": APIIntegration,
            "database": DatabaseIntegration,
        }

        logger.info(f"Registered {len(self.integration_classes)} integration types")

    async def initialize(self) -> bool:
        """
        Initialize the integration hub and configured integrations

        Returns:
            True if initialization was successful, False otherwise
        """
        try:
            logger.info("Initializing integration hub")

            # In a real implementation, this would load integration configurations
            # from settings or database

            # For demonstration, we'll create a fake example integration
            example_config = IntegrationConfig(
                enabled=True,
                name="example_api",
                type="api",
                config={
                    "base_url": "https://api.example.com/v1",
                    "api_key": "fake-api-key",
                    "timeout": 30,
                },
                metadata={
                    "description": "Example API integration",
                },
            )

            # Register and initialize example integration
            await self.register_integration(example_config)

            logger.info("Integration hub initialization complete")
            return True

        except Exception as e:
            logger.error(
                f"Failed to initialize integration hub: {str(e)}", exc_info=True
            )
            return False

    async def shutdown(self) -> None:
        """Shutdown the integration hub and all active integrations"""
        try:
            logger.info("Shutting down integration hub")

            shutdown_tasks = []
            for name, integration in self.integrations.items():
                if integration.initialized:
                    logger.info(f"Shutting down integration: {name}")
                    shutdown_tasks.append(integration.shutdown())

            # Wait for all integrations to shut down
            if shutdown_tasks:
                await asyncio.gather(*shutdown_tasks)

            # Clear integrations
            self.integrations.clear()

            logger.info("Integration hub shutdown complete")

        except Exception as e:
            logger.error(
                f"Error during integration hub shutdown: {str(e)}", exc_info=True
            )

    async def register_integration(self, config: IntegrationConfig) -> bool:
        """
        Register a new integration

        Args:
            config: Integration configuration

        Returns:
            True if registration was successful, False otherwise

        Raises:
            ValueError: If integration type is invalid or config is invalid
        """
        try:
            integration_type = config.type
            integration_name = config.name

            logger.info(
                f"Registering integration: {integration_name} ({integration_type})"
            )

            # Check if integration type is supported
            if integration_type not in self.integration_classes:
                raise ValueError(f"Unsupported integration type: {integration_type}")

            # Check if integration with this name already exists
            if integration_name in self.integrations:
                raise ValueError(
                    f"Integration with name '{integration_name}' already exists"
                )

            # Create integration instance
            integration_class = self.integration_classes[integration_type]
            integration = integration_class(config)

            # Add to integrations dict
            self.integrations[integration_name] = integration

            # Initialize if enabled
            if integration.enabled:
                success = await integration.initialize()
                if not success:
                    logger.warning(
                        f"Failed to initialize integration: {integration_name}"
                    )
                    return False

            logger.info(f"Successfully registered integration: {integration_name}")
            return True

        except Exception as e:
            logger.error(f"Failed to register integration: {str(e)}", exc_info=True)
            raise ValueError(f"Failed to register integration: {str(e)}")

    async def unregister_integration(self, name: str) -> bool:
        """
        Unregister an integration

        Args:
            name: Name of the integration to unregister

        Returns:
            True if unregistration was successful, False otherwise
        """
        try:
            if name not in self.integrations:
                logger.warning(f"Integration not found: {name}")
                return False

            integration = self.integrations[name]

            # Shutdown integration if initialized
            if integration.initialized:
                logger.info(f"Shutting down integration: {name}")
                await integration.shutdown()

            # Remove from integrations dict
            del self.integrations[name]

            logger.info(f"Unregistered integration: {name}")
            return True

        except Exception as e:
            logger.error(f"Failed to unregister integration: {str(e)}", exc_info=True)
            return False

    def get_integration(self, name: str) -> Optional[Integration]:
        """
        Get an integration by name

        Args:
            name: Name of the integration

        Returns:
            Integration instance or None if not found
        """
        return self.integrations.get(name)

    def list_integrations(self) -> List[IntegrationInfo]:
        """
        List all registered integrations

        Returns:
            List of integration information
        """
        return [integration.get_info() for integration in self.integrations.values()]

    async def execute_integration_method(
        self,
        integration_name: str,
        method_name: str,
        **kwargs: Any,
    ) -> Any:
        """
        Execute a method on an integration

        Args:
            integration_name: Name of the integration
            method_name: Name of the method to execute
            **kwargs: Method arguments

        Returns:
            Method execution result

        Raises:
            ValueError: If integration not found or method not supported
        """
        try:
            # Get integration
            integration = self.get_integration(integration_name)
            if integration is None:
                raise ValueError(f"Integration not found: {integration_name}")

            # Check if integration is enabled and initialized
            if not integration.enabled:
                raise ValueError(f"Integration is disabled: {integration_name}")

            if not integration.initialized:
                raise ValueError(f"Integration is not initialized: {integration_name}")

            # Check if method exists
            if not hasattr(integration, method_name):
                raise ValueError(
                    f"Method '{method_name}' not supported by integration '{integration_name}'"
                )

            # Get method
            method = getattr(integration, method_name)

            # Check if method is callable
            if not callable(method):
                raise ValueError(
                    f"'{method_name}' is not a callable method on integration '{integration_name}'"
                )

            # Execute method
            if asyncio.iscoroutinefunction(method):
                return await method(**kwargs)
            else:
                return method(**kwargs)

        except Exception as e:
            logger.error(
                f"Failed to execute method '{method_name}' on integration '{integration_name}': {str(e)}",
                exc_info=True,
            )
            raise ValueError(f"Failed to execute integration method: {str(e)}")


# Example API integration implementation
class ExampleAPIIntegration(APIIntegration):
    """Example API integration implementation"""

    async def initialize(self) -> bool:
        """Initialize the integration"""
        try:
            self.logger.info(f"Initializing {self.name} integration")

            # In a real implementation, this would set up HTTP clients,
            # validate API keys, etc.

            # Simulate API connection test
            connection_success = await self.test_connection()

            if connection_success:
                self.status = IntegrationStatus.ACTIVE
                self._initialized = True
                self.logger.info(f"{self.name} integration initialized successfully")
                return True
            else:
                self.status = IntegrationStatus.ERROR
                self.logger.error(f"Failed to establish connection for {self.name}")
                return False

        except Exception as e:
            self.status = IntegrationStatus.ERROR
            self.logger.error(
                f"Failed to initialize {self.name}: {str(e)}", exc_info=True
            )
            return False

    async def shutdown(self) -> None:
        """Shutdown the integration"""
        try:
            self.logger.info(f"Shutting down {self.name} integration")

            # In a real implementation, this would close HTTP clients,
            # clean up resources, etc.

            self.status = IntegrationStatus.CONFIGURED
            self._initialized = False

            self.logger.info(f"{self.name} integration shut down successfully")

        except Exception as e:
            self.logger.error(
                f"Error shutting down {self.name}: {str(e)}", exc_info=True
            )

    def get_info(self) -> IntegrationInfo:
        """Get integration information"""
        return IntegrationInfo(
            name=self.name,
            type=self.config.type,
            description=self.config.metadata.get(
                "description", "Example API integration"
            ),
            status=self.status,
            capabilities=["data_retrieval", "data_submission"],
            config_schema={
                "base_url": {"type": "string", "required": True},
                "api_key": {"type": "string", "required": True},
                "timeout": {"type": "integer", "default": 30},
            },
            metadata=self.config.metadata,
        )

    async def test_connection(self) -> bool:
        """Test the API connection"""
        try:
            self.logger.debug(f"Testing connection for {self.name}")

            # In a real implementation, this would make a simple API call
            # to verify connectivity

            # Simulate successful connection
            await asyncio.sleep(0.5)

            self.logger.debug(f"Connection test successful for {self.name}")
            return True

        except Exception as e:
            self.logger.error(
                f"Connection test failed for {self.name}: {str(e)}", exc_info=True
            )
            return False

    async def get_data(
        self, endpoint: str, params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Example method to get data from the API

        Args:
            endpoint: API endpoint path
            params: Optional query parameters

        Returns:
            API response data
        """
        if not self.initialized:
            raise RuntimeError(f"Integration {self.name} is not initialized")

        # In a real implementation, this would make an actual HTTP request

        self.logger.debug(f"Getting data from {endpoint} with params: {params}")

        # Simulate API call
        await asyncio.sleep(0.5)

        # Return mock response
        return {
            "success": True,
            "data": {
                "message": f"Data from {endpoint}",
                "timestamp": "2025-03-06T15:30:45Z",
            },
            "metadata": {
                "params": params or {},
            },
        }


# Global instance
integration_hub = None


def initialize_integration_hub() -> IntegrationHub:
    """
    Initialize the global integration hub instance

    Returns:
        The initialized integration hub
    """
    global integration_hub
    if integration_hub is None:
        integration_hub = IntegrationHub()
    return integration_hub


# Export public symbols
__all__ = [
    "IntegrationStatus",
    "IntegrationConfig",
    "IntegrationInfo",
    "Integration",
    "APIIntegration",
    "DatabaseIntegration",
    "IntegrationHub",
    "ExampleAPIIntegration",
    "integration_hub",
    "initialize_integration_hub",
]

File: cortex/node/src/components/integration-hub.ts
/**
 * Implementation of the Integration Hub component for Cortex Core
 */

import { v4 as uuidv4 } from 'uuid';
import { logger } from '../utils/logger';
import { db } from '../database/connection';
import { MCPServer, MCPClient } from '../utils/mcp';
import { EventEmitter } from 'events';
import { stringifyJson, parseJsonString, parseStringArray } from '../utils/json-helpers';
import config from '../config';

export interface Integration {
  id: string;
  name: string;
  type: "vscode" | "m365" | "browser" | "other";
  connectionDetails: ConnectionDetails;
  capabilities: string[];
  status: "connected" | "disconnected" | "error";
  lastActive: Date;
}

export interface ConnectionDetails {
  protocol: "mcp" | "rest" | "websocket";
  endpoint: string;
  authToken?: string;
  metadata?: Record<string, any>;
}

export class IntegrationHub extends EventEmitter {
  private readonly integrations: Map<string, Integration> = new Map();
  private readonly mcpClients: Map<string, MCPClient> = new Map();
  private readonly mcpServer: MCPServer;

  constructor() {
    super();

    // Initialize MCP server
    this.mcpServer = new MCPServer({
      name: 'cortex-core',
      version: '1.0.0',
      capabilities: {
        resources: {},
        tools: {},
        prompts: {}
      }
    });

    // Set up MCP server event handlers
    this.mcpServer.on('connection', this.handleMCPConnection.bind(this));
    this.mcpServer.on('disconnection', this.handleMCPDisconnection.bind(this));
  }

  /**
   * Initialize the Integration Hub
   */
  async initialize(): Promise<void> {
    try {
      logger.info('Initializing Integration Hub');

      // Start MCP server
      await this.mcpServer.start();

      // Load previously registered integrations from DB
      const storedIntegrations = await db.integrations.findMany();

      for (const integration of storedIntegrations) {
        // Parse JSON strings into proper objects
        this.integrations.set(integration.id, {
          id: integration.id,
          name: integration.name,
          type: integration.type as "vscode" | "m365" | "browser" | "other",
          connectionDetails: parseJsonString(integration.connectionDetails, {
            protocol: 'mcp',
            endpoint: ''
          }),
          capabilities: parseStringArray(integration.capabilitiesJson),
          status: 'disconnected', // All stored integrations start as disconnected
          lastActive: new Date(integration.lastActive)
        });
      }

      logger.info(`Loaded ${storedIntegrations.length} integrations from database`);

      // Register configured MCP endpoints from config
      if (config.mcp && config.mcp.endpoints && config.mcp.endpoints.length > 0) {
        logger.info(`Registering ${config.mcp.endpoints.length} configured MCP endpoints`);

        for (const endpoint of config.mcp.endpoints) {
          try {
            logger.info(`Registering MCP endpoint: ${endpoint.name} at ${endpoint.endpoint}`);

            // Check if this endpoint is already registered
            const existingIntegration = Array.from(this.integrations.values()).find(
              integration =>
                integration.connectionDetails.endpoint === endpoint.endpoint &&
                integration.name === endpoint.name
            );

            if (existingIntegration) {
              logger.info(`MCP endpoint ${endpoint.name} already registered with ID ${existingIntegration.id}`);
              continue;
            }

            // Register new integration
            await this.registerIntegration({
              name: endpoint.name,
              type: endpoint.type as "vscode" | "m365" | "browser" | "other",
              connectionDetails: {
                protocol: 'mcp',
                endpoint: endpoint.endpoint
              },
              capabilities: []
            });

            logger.info(`Successfully registered MCP endpoint: ${endpoint.name}`);
          } catch (error) {
            logger.error(`Failed to register MCP endpoint: ${endpoint.name}`, error);
          }
        }
      } else {
        logger.info('No MCP endpoints configured');
      }
    } catch (error) {
      logger.error('Failed to initialize Integration Hub', error);
      throw new Error(`Failed to initialize Integration Hub: ${(error as Error).message}`);
    }
  }

  /**
   * Register a new external integration
   * @param integration The integration to register
   */
  async registerIntegration(integration: Omit<Integration, 'id' | 'status' | 'lastActive'>): Promise<Integration> {
    try {
      logger.info(`Registering new integration: ${integration.name} (${integration.type})`);

      const newIntegration: Integration = {
        ...integration,
        id: uuidv4(),
        status: 'disconnected',
        lastActive: new Date()
      };

      // Save to database with JSON serialization
      await db.integrations.create({
        data: {
          id: newIntegration.id,
          name: newIntegration.name,
          type: newIntegration.type,
          connectionDetails: stringifyJson(newIntegration.connectionDetails),  // Convert to JSON string
          capabilitiesJson: stringifyJson(newIntegration.capabilities),        // Convert to JSON string
          lastActive: newIntegration.lastActive
        }
      });

      // Add to in-memory map
      this.integrations.set(newIntegration.id, newIntegration);

      logger.info(`Successfully registered integration with ID: ${newIntegration.id}`);
      return newIntegration;
    } catch (error) {
      logger.error(`Failed to register integration: ${integration.name}`, error);
      throw new Error(`Failed to register integration: ${(error as Error).message}`);
    }
  }

  /**
   * Get an integration by ID
   * @param integrationId The ID of the integration to retrieve
   * @returns The integration or null if not found
   */
  async getIntegration(integrationId: string): Promise<Integration | null> {
    try {
      return this.integrations.get(integrationId) || null;
    } catch (error) {
      logger.error(`Failed to get integration: ${integrationId}`, error);
      return null;
    }
  }

  /**
   * Forward a request to an external integration
   * @param integrationId The ID of the integration to forward to
   * @param request The request to forward
   * @returns The response from the integration
   */
  async forwardRequest(integrationId: string, request: any): Promise<any> {
    try {
      logger.info(`Forwarding request to integration: ${integrationId}`);

      const integration = await this.getIntegration(integrationId);
      if (!integration) {
        throw new Error(`Integration not found: ${integrationId}`);
      }

      if (integration.status !== 'connected') {
        throw new Error(`Integration is not connected: ${integrationId}`);
      }

      // Handle based on protocol
      if (integration.connectionDetails.protocol === 'mcp') {
        // Get MCP client
        const client = this.mcpClients.get(integrationId);
        if (!client) {
          throw new Error(`MCP client not found for integration: ${integrationId}`);
        }

        // Forward via MCP
        const response = await client.sendRequest(request.method, request.params);

        // Update last active timestamp
        await this.updateIntegrationLastActive(integrationId);

        return response;
      } else if (integration.connectionDetails.protocol === 'rest') {
        // Implement REST forwarding
        throw new Error('REST protocol not implemented yet');
      } else if (integration.connectionDetails.protocol === 'websocket') {
        // Implement WebSocket forwarding
        throw new Error('WebSocket protocol not implemented yet');
      } else {
        throw new Error(`Unsupported protocol: ${integration.connectionDetails.protocol}`);
      }
    } catch (error) {
      logger.error(`Failed to forward request to integration: ${integrationId}`, error);
      throw new Error(`Failed to forward request: ${(error as Error).message}`);
    }
  }

  /**
   * Handle incoming requests from external integrations
   * @param integrationId The ID of the integration sending the request
   * @param request The incoming request
   * @returns The response to send back
   */
  async handleExternalRequest(integrationId: string, request: any): Promise<any> {
    try {
      logger.info(`Handling external request from integration: ${integrationId}`);

      const integration = await this.getIntegration(integrationId);
      if (!integration) {
        throw new Error(`Integration not found: ${integrationId}`);
      }

      // Update last active timestamp
      await this.updateIntegrationLastActive(integrationId);

      // Emit an event for this request
      this.emit('externalRequest', {
        integrationId,
        request,
        integration
      });

      // For now, return a simple acknowledgment
      // In a real implementation, this would route the request to the appropriate handler
      return {
        success: true,
        message: 'Request received'
      };
    } catch (error) {
      logger.error(`Failed to handle external request from integration: ${integrationId}`, error);
      return {
        success: false,
        error: (error as Error).message || 'Unknown error'
      };
    }
  }

  /**
   * List all active integrations
   * @returns Array of registered integrations
   */
  async listIntegrations(): Promise<Integration[]> {
    try {
      return Array.from(this.integrations.values());
    } catch (error) {
      logger.error('Failed to list integrations', error);
      return [];
    }
  }

  /**
   * Handle new MCP connection
   * @param connection The MCP connection details
   * @private
   */
  private async handleMCPConnection(connection: any): Promise<void> {
    try {
      logger.info(`New MCP connection from: ${connection.clientName}`);

      // Find if this is an existing integration by endpoint or create a new one
      let integration = Array.from(this.integrations.values()).find(
        i => i.connectionDetails.endpoint === connection.endpoint
      );

      if (!integration) {
        // This is a new integration, register it
        integration = await this.registerIntegration({
          name: connection.clientName,
          type: 'other', // Default type
          connectionDetails: {
            protocol: 'mcp',
            endpoint: connection.endpoint
          },
          capabilities: []
        });
      }

      // Update status to connected
      integration.status = 'connected';
      await this.updateIntegration(integration.id, { status: 'connected' });

      // Create and store MCP client
      const client = new MCPClient(connection);
      this.mcpClients.set(integration.id, client);

      // Set up client event handlers
      client.on('request', async (req) => {
        const response = await this.handleExternalRequest(integration.id, req);
        return response;
      });

      client.on('disconnected', () => {
        this.handleMCPDisconnection(connection);
      });

      // Emit connection event
      this.emit('integrationConnected', integration);
    } catch (error) {
      logger.error('Failed to handle MCP connection', error);
    }
  }

  /**
   * Handle MCP disconnection
   * @param connection The MCP connection details
   * @private
   */
  private async handleMCPDisconnection(connection: any): Promise<void> {
    try {
      logger.info(`MCP disconnection from: ${connection.clientName}`);

      // Find the integration
      const integration = Array.from(this.integrations.values()).find(
        i => i.connectionDetails.endpoint === connection.endpoint
      );

      if (integration) {
        // Update status to disconnected
        integration.status = 'disconnected';
        await this.updateIntegration(integration.id, { status: 'disconnected' });

        // Remove MCP client
        this.mcpClients.delete(integration.id);

        // Emit disconnection event
        this.emit('integrationDisconnected', integration);
      }
    } catch (error) {
      logger.error('Failed to handle MCP disconnection', error);
    }
  }

  /**
   * Update integration details
   * @param integrationId The ID of the integration to update
   * @param updates The updates to apply
   * @private
   */
  private async updateIntegration(integrationId: string, updates: Partial<Integration>): Promise<void> {
    try {
      const integration = this.integrations.get(integrationId);

      if (integration) {
        // Apply updates to in-memory integration
        Object.assign(integration, updates);

        // Prepare database updates with JSON serialization
        const dbUpdates: any = { ...updates };

        // Convert objects to JSON strings for database storage
        if (updates.connectionDetails) {
          dbUpdates.connectionDetails = stringifyJson(updates.connectionDetails);
        }

        if (updates.capabilities) {
          dbUpdates.capabilitiesJson = stringifyJson(updates.capabilities);
          delete dbUpdates.capabilities; // Remove the original object
        }

        // Update in database
        await db.integrations.update({
          where: { id: integrationId },
          data: dbUpdates
        });
      }
    } catch (error) {
      logger.error(`Failed to update integration: ${integrationId}`, error);
    }
  }

  /**
   * Update the last active timestamp for an integration
   * @param integrationId The ID of the integration
   * @private
   */
  private async updateIntegrationLastActive(integrationId: string): Promise<void> {
    const now = new Date();
    await this.updateIntegration(integrationId, { lastActive: now });
  }
}
